### ISSUE TITLE
Pushing over SSH

### ISSUE THREAD

**Issue by datashard** [icyphox/legit#47]
More of a question than an actual Issue, because I'm kinda lost.

How do I get pushing over SSH to work when using the Docker Image / docker compose?

**Comment by icyphox** [comment]
I’m not entirely sure how you’re running the container, but in my K3s homelab I have my git repos on the host itself, and I push to it using the “git” user with a `.ssh/authorized_keys` file in its home directory. I then mount this home directory inside my legit container. Not sure if that helps.

**Comment by datashard** [comment]
It does! This confirms my earlier thought lmao

Guessing the `.ssh/authorized_keys` file would be in `/home/git/.ssh/authorized_keys` within the container?

**Comment by icyphox** [comment]
Hm, I haven't tried running ssh in the container since I push to the underlying host's `/home/git` (container has read-only access to this dir) but I suppose that could work as well.

**Comment by datashard** [comment]
I'll try it! :D 

I'll report back if it works

**Comment by datashard** [comment]
Well, ssh-ing into the container doesn't work. Shouldn't be surprised since I doubt there's a SSH Server running in there? It doesn't even have (ba)sh so I should've figured that out sooner lol

Would you mind sharing the code for the way you push?
If I can't get it to work I might just run an hourly cron to `cp` certain things from my forgejo instance.

As for general http auth, you could theoretically try and use `htpasswd`? I haven't looked into it much but I think it should work?

**Comment by icyphox** [comment]
Oh, yeah, of course—there is no ssh inside the container. It's a very light image.

There isn't much "code" to share. I have the git repos on a machine and push over ssh as you would anywhere else. This directory of bare git repositories gets mounted inside my legit container. legit itself doesn't deal with any of the ssh pushing.

As for http auth, that isn't supported by legit (explicitly disabled). The recommended approach is what I described above. Bear in mind: legit is read-only. It will simply serve git repos in available at `repo.scanPath`. Populating that directory is the user's problem.

**Comment by datashard** [comment]
Trying to actually get this working on my end again, and I was wondering how you're setting up the repos on your end?

I've been trying to use `git init --bare` and pushing to it after, and while the repo exists on the server, I get a `reference not found` error in the logs and though when i copy over a repo from my forgejo instance, those just show up without problems.

**Comment by datashard** [comment]
Yeah. I'm really lost about this and have been pulling my hair out trying to figure out what the difference between repos is. because even after pushing to the remote repo (made with git init --bare) it did not show up. 

All I get from my Logs is `getting head of /var/www/git/bare.git: reference not found`, which doesn't really give me anything to figure out _what_ I'm missing in the repo

**Comment by icyphox** [comment]
Hey! Hm, I can try and help you out on something quicker than here, if you prefer. Have you got Discord/email/…?

**Comment by icyphox** [comment]
But yeah, that's not really a "breaking" error per se. That's just legit complaining that it can't get the HEAD (the topmost ref, or commit, if you will). And it can't get it because there isn't any—a bare repo has no commits. So once you push a commit, it'll show up and the log message will go away.

**Comment by datashard** [comment]
> Hey! Hm, I can try and help you out on something quicker than here, if you prefer. Have you got Discord/email/…?

I have Discord `@fangmarks`

> But yeah, that's not really a "breaking" error per se. That's just legit complaining that it can't get the HEAD—the topmost ref, or commit, if you will—of the repository. And it can't get it because there isn't any: a newly initialized bare repo _has no_ commits. So once you push a commit, it'll show up and the log message will go away.

That's the thing though, even after I pushed a commit, it still wasn't showing up

**Comment by datashard** [comment]
I... have no idea what the fuck I did but now it's showing up????

**Comment by icyphox** [comment]
lol, strange. Possibly wonky browser cache...? I can't be sure.

**Comment by runxiyu** [comment]
I've been trying to implement protocols natively (i.e., SSH directly handled in the forge daemon, without relying on OpenSSH or git-upload-pack/git-receive-pack) in my own Git forge for quite a while and it's a huge pain. I find go-git's documentation on network transports to be rather lacking.

* There's an old fork of legit that [says it implements ssh pushing](https://git.mills.io/prologic/legit/commit/6f8b2b27f5c3d35dfb41a5a38a3795d8c4962692), I'm currently auditing it to see if it works
* Someone has [an example of cloning via ssh, with go-git](https://github.com/erred/gitreposerver/blob/main/ssh.go)
* SSH pushing isn't implemented at all in my git forge, where anonymous SSH cloning still [spawns an external git-upload-pack process](https://forge.lindenii.runxiyu.org/lindenii/:/repos/forge/tree/ssh_server.go?commit=1159d89c4dcd11059cffda004bb64be648c7918d)

I suppose this is out of scope for legit? Though anonymous SSH cloning would definitely be nice (no need to trust PKI/CAs, etc.).

**Comment by icyphox** [comment]
@runxiyu We used to have pure-Go git-upload-pack implementation, see: [v0.2.2](https://github.com/icyphox/legit/tree/v0.2.2) tree. This was changed in v0.2.3 due to an [upstream bug](https://github.com/go-git/go-git/issues/1062), which has now since been fixed. But! The performance of go-git's upload-pack leaves a lot to be desired (see my comment in the linked issue)—so I've stuck to the system's git-upload-pack binary for now. 

Pure Go SSH sounds very interesting, I'll admit.  Anonymous SSH cloning would certainly be very nice. I'm happy to consider a patch if you'd like to try implementing this!

**Comment by runxiyu** [comment]
@icyphox Although that stack overflow has been solved as far as I know, I still get various SSH failures related to [upstream #647](https://github.com/go-git/go-git/issues/647) when implementing it myself... so I switched cloning/pushing over both HTTPS and SSH to using system git (which was annoying because now I have to implement hook handlers and stuff). If I get the time I'll prepare a patchset for legit to support anonymous SSH cloning :)
